Create RemoteScrollingCoordinatorProxyIOS

smfr opened the pull request:
#### e618b68f1b6a688be17db4c3b2b23de8f940bf45
<pre>
Create RemoteScrollingCoordinatorProxyIOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=247518">https://bugs.webkit.org/show_bug.cgi?id=247518</a>
rdar://101987258

Reviewed by Sam Weinig.

macOS subclassed RemoteScrollingCoordinatorProxy into RemoteScrollingCoordinatorProxyMac, so do the same
for iOS. This cleans up some #ifdeffed code.

The RemoteLayerTreeDrawingAreaProxy subclasses instantiate the correct type of RemoteScrollingCoordinatorProxy.

Define SPECIALIZE_TYPE_TRAITS so we can use downcast&lt;&gt; etc but have the predicate be compile-time.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView scrollViewWillBeginDragging:]):
(-[WKWebView scrollViewWillEndDragging:withVelocity:targetContentOffset:]):
(-[WKWebView _updateVisibleContentRects]):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::createScrollingCoordinatorProxy const): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::layerTreeHost const):
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeDidScroll):
(WebKit::RemoteScrollingCoordinatorProxy::connectStateNodeLayers): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::establishLayerTreeScrollingRelations): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::activeTouchActionsForTouchIdentifier const): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::setTouchActionsForTouchIdentifier): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::clearTouchActionsForTouchIdentifier): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
(WebKit::RemoteScrollingCoordinatorProxy::isRemoteScrollingCoordinatorProxyIOS const):
(WebKit::RemoteScrollingCoordinatorProxy::isRemoteScrollingCoordinatorProxyMac const):
(WebKit::RemoteScrollingCoordinatorProxy::propagatesMainFrameScrolls const):
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeWillStartPanGesture):
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeWillStartScroll):
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeDidEndScroll):
(WebKit::RemoteScrollingCoordinatorProxy::setRootNodeIsInUserScroll):
(WebKit::RemoteScrollingCoordinatorProxy::setPropagatesMainFrameScrolls): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::fixedScrollingNodeLayerIDs const): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeDrawingAreaProxyIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeDrawingAreaProxyIOS.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyIOS::createScrollingCoordinatorProxy const):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.h: Added.
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::RemoteScrollingCoordinatorProxyIOS):
(WebKit::RemoteScrollingCoordinatorProxyIOS::activeTouchActionsForTouchIdentifier const):
(WebKit::RemoteScrollingCoordinatorProxyIOS::setTouchActionsForTouchIdentifier):
(WebKit::RemoteScrollingCoordinatorProxyIOS::clearTouchActionsForTouchIdentifier):
(WebKit::RemoteScrollingCoordinatorProxyIOS::scrollViewForScrollingNodeID const):
(WebKit::RemoteScrollingCoordinatorProxyIOS::removeFixedScrollingNodeLayerIDs):
(WebKit::RemoteScrollingCoordinatorProxyIOS::connectStateNodeLayers):
(WebKit::RemoteScrollingCoordinatorProxyIOS::currentLayoutViewport const):
(WebKit::RemoteScrollingCoordinatorProxyIOS::scrollingTreeNodeWillStartPanGesture):
(WebKit::RemoteScrollingCoordinatorProxyIOS::scrollingTreeNodeWillStartScroll):
(WebKit::RemoteScrollingCoordinatorProxyIOS::scrollingTreeNodeDidEndScroll):
(WebKit::RemoteScrollingCoordinatorProxyIOS::establishLayerTreeScrollingRelations):
(WebKit::RemoteScrollingCoordinatorProxyIOS::adjustTargetContentOffsetForSnapping):
(WebKit::RemoteScrollingCoordinatorProxyIOS::shouldSetScrollViewDecelerationRateFast const):
(WebKit::RemoteScrollingCoordinatorProxyIOS::setRootNodeIsInUserScroll):
(WebKit::RemoteScrollingCoordinatorProxyIOS::shouldSnapForMainFrameScrolling const):
(WebKit::RemoteScrollingCoordinatorProxyIOS::closestSnapOffsetForMainFrameScrolling const):
(WebKit::RemoteScrollingCoordinatorProxyIOS::hasActiveSnapPoint const):
(WebKit::RemoteScrollingCoordinatorProxyIOS::nearestActiveContentInsetAdjustedSnapOffset const):
(WebKit::RemoteScrollingCoordinatorProxy::scrollViewForScrollingNodeID const): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::removeFixedScrollingNodeLayerIDs): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::connectStateNodeLayers): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::currentLayoutViewport const): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeWillStartPanGesture): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeWillStartScroll): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeDidEndScroll): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::establishLayerTreeScrollingRelations): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::adjustTargetContentOffsetForSnapping): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::shouldSetScrollViewDecelerationRateFast const): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::setRootNodeIsInUserScroll): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::shouldSnapForMainFrameScrolling const): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::closestSnapOffsetForMainFrameScrolling const): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::hasActiveSnapPoint const): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::nearestActiveContentInsetAdjustedSnapOffset const): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.mm:
(WebKit::ScrollingTreeScrollingNodeDelegateIOS::computeActiveTouchActionsForGestureRecognizer):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::connectStateNodeLayers):
(WebKit::RemoteScrollingCoordinatorProxyMac::establishLayerTreeScrollingRelations):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setDrawingArea):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _handleTouchActionsForTouchEvent:]):
(-[WKContentView _scrollingNodeScrollingDidEnd:]):
(-[WKContentView _updateTargetedPreviewScrollViewUsingContainerScrollingNodeID:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/256372@main">https://commits.webkit.org/256372@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9dc687611571f354f9fea2c2afe27d5230560009

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95565 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4829 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28613 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105145 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4866 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33571 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87936 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100996 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101227 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3554 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82177 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30636 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85441 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87354 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73473 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39317 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18908 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37017 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20210 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4403 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/41014 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42858 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/43002 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39461 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
```diff
@@ -51,9 +50,32 @@
+    auto iterator = m_touchActionsByTouchIdentifier.find(touchIdentifier);
+    if (iterator == m_touchActionsByTouchIdentifier.end())
+        return { };
+    return iterator->value;
```
Would `m_touchActionsByTouchIdentifier.get(touchIdentifier);` work here? I think the return value of `.get()` if the identifier is not found should be an empty OptionSet.

weinig commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
```diff
@@ -227,14 +249,14 @@
+void RemoteScrollingCoordinatorProxyIOS::setRootNodeIsInUserScroll(bool value)
 {
-    ScrollingTreeNode* root = m_scrollingTree->rootNode();
+    ScrollingTreeNode* root = scrollingTree()->rootNode();
```
You could change a few of these to `auto*` if you so desired.

smfr committed:
Create RemoteScrollingCoordinatorProxyIOS
https://bugs.webkit.org/show_bug.cgi?id=247518
rdar://101987258

Reviewed by Sam Weinig.

macOS subclassed RemoteScrollingCoordinatorProxy into RemoteScrollingCoordinatorProxyMac, so do the same
for iOS. This cleans up some #ifdeffed code.

The RemoteLayerTreeDrawingAreaProxy subclasses instantiate the correct type of RemoteScrollingCoordinatorProxy.

Define SPECIALIZE_TYPE_TRAITS so we can use downcast<> etc but have the predicate be compile-time.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView scrollViewWillBeginDragging:]):
(-[WKWebView scrollViewWillEndDragging:withVelocity:targetContentOffset:]):
(-[WKWebView _updateVisibleContentRects]):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::createScrollingCoordinatorProxy const): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::layerTreeHost const):
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeDidScroll):
(WebKit::RemoteScrollingCoordinatorProxy::connectStateNodeLayers): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::establishLayerTreeScrollingRelations): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::activeTouchActionsForTouchIdentifier const): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::setTouchActionsForTouchIdentifier): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::clearTouchActionsForTouchIdentifier): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
(WebKit::RemoteScrollingCoordinatorProxy::isRemoteScrollingCoordinatorProxyIOS const):
(WebKit::RemoteScrollingCoordinatorProxy::isRemoteScrollingCoordinatorProxyMac const):
(WebKit::RemoteScrollingCoordinatorProxy::propagatesMainFrameScrolls const):
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeWillStartPanGesture):
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeWillStartScroll):
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeDidEndScroll):
(WebKit::RemoteScrollingCoordinatorProxy::setRootNodeIsInUserScroll):
(WebKit::RemoteScrollingCoordinatorProxy::setPropagatesMainFrameScrolls): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::fixedScrollingNodeLayerIDs const): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeDrawingAreaProxyIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeDrawingAreaProxyIOS.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyIOS::createScrollingCoordinatorProxy const):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.h: Added.
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::RemoteScrollingCoordinatorProxyIOS):
(WebKit::RemoteScrollingCoordinatorProxyIOS::activeTouchActionsForTouchIdentifier const):
(WebKit::RemoteScrollingCoordinatorProxyIOS::setTouchActionsForTouchIdentifier):
(WebKit::RemoteScrollingCoordinatorProxyIOS::clearTouchActionsForTouchIdentifier):
(WebKit::RemoteScrollingCoordinatorProxyIOS::scrollViewForScrollingNodeID const):
(WebKit::RemoteScrollingCoordinatorProxyIOS::removeFixedScrollingNodeLayerIDs):
(WebKit::RemoteScrollingCoordinatorProxyIOS::connectStateNodeLayers):
(WebKit::RemoteScrollingCoordinatorProxyIOS::currentLayoutViewport const):
(WebKit::RemoteScrollingCoordinatorProxyIOS::scrollingTreeNodeWillStartPanGesture):
(WebKit::RemoteScrollingCoordinatorProxyIOS::scrollingTreeNodeWillStartScroll):
(WebKit::RemoteScrollingCoordinatorProxyIOS::scrollingTreeNodeDidEndScroll):
(WebKit::RemoteScrollingCoordinatorProxyIOS::establishLayerTreeScrollingRelations):
(WebKit::RemoteScrollingCoordinatorProxyIOS::adjustTargetContentOffsetForSnapping):
(WebKit::RemoteScrollingCoordinatorProxyIOS::shouldSetScrollViewDecelerationRateFast const):
(WebKit::RemoteScrollingCoordinatorProxyIOS::setRootNodeIsInUserScroll):
(WebKit::RemoteScrollingCoordinatorProxyIOS::shouldSnapForMainFrameScrolling const):
(WebKit::RemoteScrollingCoordinatorProxyIOS::closestSnapOffsetForMainFrameScrolling const):
(WebKit::RemoteScrollingCoordinatorProxyIOS::hasActiveSnapPoint const):
(WebKit::RemoteScrollingCoordinatorProxyIOS::nearestActiveContentInsetAdjustedSnapOffset const):
(WebKit::RemoteScrollingCoordinatorProxy::scrollViewForScrollingNodeID const): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::removeFixedScrollingNodeLayerIDs): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::connectStateNodeLayers): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::currentLayoutViewport const): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeWillStartPanGesture): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeWillStartScroll): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeDidEndScroll): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::establishLayerTreeScrollingRelations): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::adjustTargetContentOffsetForSnapping): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::shouldSetScrollViewDecelerationRateFast const): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::setRootNodeIsInUserScroll): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::shouldSnapForMainFrameScrolling const): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::closestSnapOffsetForMainFrameScrolling const): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::hasActiveSnapPoint const): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::nearestActiveContentInsetAdjustedSnapOffset const): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.mm:
(WebKit::ScrollingTreeScrollingNodeDelegateIOS::computeActiveTouchActionsForGestureRecognizer):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::connectStateNodeLayers):
(WebKit::RemoteScrollingCoordinatorProxyMac::establishLayerTreeScrollingRelations):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setDrawingArea):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _handleTouchActionsForTouchEvent:]):
(-[WKContentView _scrollingNodeScrollingDidEnd:]):
(-[WKContentView _updateTargetedPreviewScrollViewUsingContainerScrollingNodeID:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/256372@main
